cupy/linalg/einsum.py

```diff
@@ -0,0 +1,251 @@
+            self.result = self.ioperand
+        for label in self.label_to_summed:
+            self.subscript = self.subscript.replace(label, '')
+
+
+class TransposedViewCalculator(object):
+    """Calculate 'ij->ji' by cupy.transpose if needed.
+
+    Args:
+        ioperand (cupy.ndarray): Array to be transpose.
+        input_subscript (str): Specifies the subscripts for input arrays.
+        output_subscript (str):
+            Specifies the subscripts for output arrays. If input does not
+            match output, ``operand`` is transposed so that it matches.
+    """
+
+    def __init__(self, ioperand, input_subscript, output_subscript):
+        assert len(input_subscript) == len(output_subscript)
+        assert set(input_subscript) == set(output_subscript)
+        self.ioperand = ioperand
+        self.input_subscript = input_subscript
+        self.output_subscript = output_subscript
+
+    def __call__(self):
+        transpose_orders = []
+        for label in self.output_subscript:
+            transpose_orders.append(self.input_subscript.find(label))
+        if transpose_orders == sorted(transpose_orders):
+            self.result = self.ioperand
+        else:
+            self.result = self.ioperand.transpose(transpose_orders)
+
+
+class CombinedViewCalculator(object):
+    """Calculate 'i,j->ij' by cupy.tensordot.
+
+    Args:
+        ioperands (sequence of arrays): Arrays to be combined.
+        subscripts (sequence of str): Specifies the subscripts.
+    """
+
+    def __init__(self, ioperands, subscripts):
+        self.ioperands = ioperands
+        self.subscripts = subscripts
+
+    def __call__(self):
+        self.result = self.ioperands[0]
+        for ioperand in self.ioperands[1:]:
+            # TODO(fukatani): add up at here if enable.
+            self.result = cupy.tensordot(self.result, ioperand, axes=0)
+        self.subscript = ''.join(self.subscripts)
+
+
+def get_dummy_labels(label_list):
+    dummy_label_set = set([])
+    count_dict = collections.Counter(label_list)
+    for label, count in count_dict.items():
+        if count >= 2:
+            dummy_label_set.add(label)
+    return dummy_label_set
+
+
+def einsum(*operands):
+    """Evaluates the Einstein summation convention on the operands.
+
+    Using the Einstein summation convention, many common multi-dimensional
+    array operations can be represented in a simple fashion.  This function
+    provides a way to compute such summations.
+
+    Args:
+        subscripts (str): Specifies the subscripts for summation.
+        operands (sequence of arrays): These are the arrays for the operation.
+
+    Returns:
+        cupy.ndarray:
+            The calculation based on the Einstein summation convention.
+
+    .. seealso:: :func:`numpy.einsum`
+    """
+
+    # TODO(fukatani): Support 'out', 'order', 'dtype', 'casting', 'optimize'
+
+    if not operands:
+        raise ValueError("must specify the einstein sum subscripts string and "
```

Comment:
If we use `def einsum(subscripts, *operands):`, `cupy.einsum()` raises `TypeError` instead of `ValueError`.

`numpy.einsum` also uses `def einsum(*operands):`.
See 
https://github.com/numpy/numpy/blob/e3b5e88ae6c0b546dd63b869f32427553ad36e57/numpy/core/einsumfunc.py#L813